### ISSUE TITLE
Dlc9 with cle-215

### ISSUE THREAD

**Issue by vladoshub** [enjoy-digital/litex#1266]
hello. Can anyone tell me if i can use DLC9 to flash my acorn board?

**Comment by enjoy-digital** [comment]
Hi @vladoshub,

IIRC I've been able to reprogram the Acorn through a JTAG HS2 and Vivado Hardware Manager, so I assume it should also be possible with a DLC9 cable. Can you give it a try? The SPI Flash on the Acorn is a S25FL128.

**Comment by vladoshub** [comment]
> Hi @vladoshub,
> 
> IIRC I've been able to reprogram the Acorn through a JTAG HS2 and Vivado Hardware Manager, so I assume it should also be possible with a DLC9 cable. Can you give it a try? The SPI Flash on the Acorn is a S25FL128.

ok, as I understand it, you just need to build the project, and take the files and use vivado for flash. Can you tell me what files I need for this, except for the .bin file?

**Comment by enjoy-digital** [comment]
Yes, you can just build the Acorn target with `--build`, then you'll have the files in `build/sqrl_acorn/gateware`. Open Vivado Hardware Manager, first try to load the `.bit `in SRAM and then you could try to Flash the `.bin` in the SPI Flash.

**Comment by vladoshub** [comment]
> Yes, you can just build the Acorn target with `--build`, then you'll have the files in `build/sqrl_acorn/gateware`. Open Vivado Hardware Manager, first try to load the `.bit` in SRAM and then you could try to Flash the `.bin` in the SPI Flash.

can I flash with pcie in the future? I just saw that there is also a file that indicates the range of flashing in memory, or will a .bin file for spi be enough and in the future I can use the command

`./litepcie_util flash_write acorn_pcie_operational.bin 0x400000`

**Comment by enjoy-digital** [comment]
To be able to flash over PCIe, you first need the LiteX bitstream to be loaded, compiled with LitePCIe (--with-pcie), the board installed in a PCIe port and LitePCIe kernel driver loaded.

**Comment by vladoshub** [comment]
> To be able to flash over PCIe, you first need the LiteX bitstream to be loaded, compiled with LitePCIe (--with-pcie), the board installed in a PCIe port and LitePCIe kernel driver loaded.

Thanks for the help, surprisingly everything went smoothly, I flashed SRAM and then SPI, except that I have S25FL256. Unfortunately, I did not find a JTAG cable for sale and I had to solder the wires directly to the connector. Can you tell me how fallback and operational bitstream work, let's say I want to change the design and play around with leds first, and what should I not do so that I can always use pcie? as I understand it, I won’t be able to flash the board if I change the pcie bridge, and in case of an error
 boot the backup part that is flashed into spi for the very first time

**Comment by enjoy-digital** [comment]
Hi @vladoshub,

you should be able to find the information here: https://github.com/enjoy-digital/litex/wiki/Use-LiteX-on-the-Acorn-CLE-215#acorn-giveaway. You can also read: https://docs.xilinx.com/v/u/en-US/xapp1247-multiboot-spi.

**Comment by enjoy-digital** [comment]
This should now be covered.

For first programming, using Vivado Hardware Manager with a Xilinx cable is fine: build the target, then use the generated files in `build/sqrl_acorn/gateware` (`.bit` for SRAM load, `.bin` for SPI  flash programming).

Current LiteX-Boards also no longer defaults to OpenOCD for the Acorn target: OpenFPGALoader is now the default programmer backend (`litex-hub/litex-boards@041c1607`).

For PCIe updates, the important point is to keep a known-good fallback image at offset `0x0` and only update the operational image at `0x400000`, as documented here: https://github.com/enjoy-digital/litex/wiki/Use-LiteX-on-the-Acorn-CLE-215#acorn-giveaway

So the operational update flow is:

  ```sh
  ./litepcie_util flash_write acorn_pcie_operational.bin 0x400000
  ./litepcie_util flash_reload
  ```

Since the original programming question was resolved and the multiboot/PCIe update flow is documented, I think we can close this.
